internal/logger/logger.go

```diff
@@ -230,46 +233,23 @@ func selectComponentLevels(componentLevels map[Component]Level) map[Component]Le
 	return selected
 }
 
-// truncate will truncate a string to the given width, appending "..." to the
-// end of the string if it is truncated. This routine is safe for multi-byte
-// characters.
-func truncate(str string, width uint) string {
-	if width == 0 {
-		return ""
-	}
-
-	if len(str) <= int(width) {
-		return str
-	}
-
-	// Truncate the byte slice of the string to the given width.
-	newStr := str[:width]
-
-	// Check if the last byte is at the beginning of a multi-byte character.
-	// If it is, then remove the last byte.
-	if newStr[len(newStr)-1]&0xC0 == 0xC0 {
-		return newStr[:len(newStr)-1] + TruncationSuffix
-	}
-
-	// Check if the last byte is in the middle of a multi-byte character. If
-	// it is, then step back until we find the beginning of the character.
-	if newStr[len(newStr)-1]&0xC0 == 0x80 {
-		for i := len(newStr) - 1; i >= 0; i-- {
-			if newStr[i]&0xC0 == 0xC0 {
-				return newStr[:i] + TruncationSuffix
-			}
-		}
+// FormatDocument formats a BSON document or RawValue for logging. The document is truncated
+// to the given width.
+func FormatDocument(msg bson.Raw, width uint) string {
+	if len(msg) == 0 {
+		return "{}"
 	}
 
-	return newStr + TruncationSuffix
+	return bsoncore.Document(msg).StringN(int(width)) + TruncationSuffix
 }
 
-// FormatMessage formats a BSON document for logging. The document is truncated
+// FormatString formats a String for logging. The string is truncated
 // to the given width.
-func FormatMessage(msg string, width uint) string {
-	if len(msg) == 0 {
-		return "{}"
+func FormatString(str string, width uint) string {
+	value := bsoncore.Value{
+		Type: bsoncore.Type(bson.TypeString),
+		Data: bsoncore.AppendString(nil, str),
```

Comment:
Building a `bsoncore.Value` to access the truncation logic is somewhat confusing. A more obvious pattern would be to move `truncate` to an internal package and export it so it can be used here and in `bsoncore`.

Reply:
I propose adding a new internal package called `bsoncoreutil` that exports `Truncate()`. I'm concerned that a more general name such as `strutil` will indicate there is some gap in the standard `strings` library.